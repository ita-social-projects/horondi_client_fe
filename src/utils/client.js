import { ApolloClient, gql, split, ApolloLink, InMemoryCache } from '@apollo/client';
import { WebSocketLink } from '@apollo/client/link/ws';
import { setContext } from '@apollo/client/link/context';
import fetch from 'unfetch';
import { IntrospectionFragmentMatcher } from 'apollo-cache-inmemory';
import { createUploadLink } from 'apollo-upload-client/public';
import { getMainDefinition } from '@apollo/client/utilities';

import { AUTH_ERRORS, USER_TOKENS } from '../configs';
import { getFromLocalStorage, setToLocalStorage } from '../services/local-storage.service';
import refreshAuthToken from './regenerateAuthTokenPair';

const introspectionResult = require('../fragmentTypes');

const { ACCESS_TOKEN } = USER_TOKENS;
const { ACCESS_TOKEN_IS_NOT_VALID } = AUTH_ERRORS;
const fragmentMatcher = new IntrospectionFragmentMatcher({
  introspectionQueryResultData: introspectionResult
});

export const REACT_APP_API_URL =
  window.env && window.env.REACT_APP_API_URL
    ? window.env.REACT_APP_API_URL
    : process.env.REACT_APP_API_URL;

const wsLink = new WebSocketLink({
  uri: process.env.REACT_APP_WS_API_URL,
  options: {
    reconnect: true,
    connectionParams: {
      authToken: getFromLocalStorage(ACCESS_TOKEN)
    }
  }
});

const authLink = setContext((_, { headers }) => ({
  headers: {
    ...headers,
    token: getFromLocalStorage(ACCESS_TOKEN) || ''
  }
}));

const terminatingLink = split(
  ({ query }) => {
    const { kind, operation } = getMainDefinition(query);
    return kind === 'OperationDefinition' && operation === 'subscription';
  },
  wsLink,
  authLink.concat(createUploadLink({ uri: `${REACT_APP_API_URL}/graphql` }))
);

export const client = new ApolloClient({
  fetch,
  link: ApolloLink.from([terminatingLink]),
  cache: new InMemoryCache({
    addTypename: false,
    fragmentMatcher
  })
});

export const getItems = async (query, variables = {}) => {
  try {
    const token = getFromLocalStorage(ACCESS_TOKEN);
    const queryResult = await client.query({
      query: gql`
        ${query}
      `,
      variables,
      context: {
        headers: {
          token
        }
      },
      fetchPolicy: 'no-cache'
    });

    if (queryResult.data && Object.values(queryResult.data)[0]?.message) {
      const message = Object.values(queryResult.data)[0]?.message;
      if (message === ACCESS_TOKEN_IS_NOT_VALID) {
        const tokenResult = await refreshAuthToken();
        if (tokenResult) {
          return await getItems(query, variables);
        }

        setToLocalStorage(ACCESS_TOKEN, null);
      }

      throw new Error(message);
    }
    return queryResult;
  } catch (e) {
    throw new Error(e.message);
  }
};

export const setItems = async (query, variables) => {
  try {
    const token = getFromLocalStorage(ACCESS_TOKEN);
    const mutationResult = await client.mutate({
      mutation: gql`
        ${query}
      `,
      variables,
      context: {
        headers: {
          token
        }
      },
      fetchPolicy: 'no-cache'
    });

    if (mutationResult.data && Object.values(mutationResult.data)[0]?.message) {
      const message = Object.values(mutationResult.data)[0]?.message;
      if (message === ACCESS_TOKEN_IS_NOT_VALID) {
        const tokenResult = await refreshAuthToken();
        if (tokenResult) {
          return await setItems(query, variables);
        }

        setToLocalStorage(ACCESS_TOKEN, null);
      }

      throw new Error(message);
    }
    return mutationResult;
  } catch (e) {
    throw new Error(e.message);
  }
};

export const setChatMail = (query, variables) =>
  client.mutate({
    mutation: gql`
      ${query}
    `,
    variables
  });
