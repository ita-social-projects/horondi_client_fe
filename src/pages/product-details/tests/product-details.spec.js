import React from 'react';
import { useQuery } from '@apollo/client';

import ProductDetails from '../index';

const mockUseContext = jest.fn().mockImplementation(() => ({
  isLight: true
}));

React.useContext = mockUseContext;

const useQueryData = {
  loading: false,
  error: false
};

jest.mock('../product-submit/product-submit.styles', () => ({
  useStyles: () => ({})
}));
jest.mock('../../../components/app/app.styles', () => ({
  useAppStyles: () => ({})
}));
jest.mock('../../../hooks/use-add-product-to-wishlist-handler', () => ({
  __esModule: true,
  default: () => [true, () => null]
}));
jest.mock('react-redux', () => ({
  ...jest.requireActual('react-redux'),
  useSelector: () => ({
    currency: 0,
    isLightTheme: true,
    productToSend: { options: { size: '' } }
  }),
  useDispatch: () => () => null
}));
jest.mock('@apollo/client');
jest.mock('../product-details.styles', () => ({
  useStyles: () => ({ container: '' })
}));

jest.mock('@material-ui/styles', () => ({
  ...jest.requireActual('@material-ui/styles'),
  useTheme: () => ({
    palette: {
      type: 'light'
    }
  })
}));

describe('Product details test', () => {
  let wrapper;

  it('Should render product details component', () => {
    useQuery.mockImplementation(() => ({
      ...useQueryData
    }));

    wrapper = shallow(<ProductDetails match={{ params: { id: 1 } }} />);

    expect(wrapper).toMatchSnapshot();
  });

  it('should cover other branches', () => {
    useQuery.mockImplementation(() => ({
      ...useQueryData,
      loading: true
    }));

    wrapper = shallow(<ProductDetails match={{ params: { id: 1 } }} />);
  });
});
