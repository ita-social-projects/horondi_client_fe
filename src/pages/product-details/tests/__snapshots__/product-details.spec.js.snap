// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Product details test Should render product details component 1`] = `
<div>
  <WithStyles(ForwardRef(Card))
    className="undefined  "
  >
    <ProductPath />
    <WithStyles(ForwardRef(Button))
      onClick={[Function]}
    >
      <ArrowIcon />
    </WithStyles(ForwardRef(Button))>
    <div>
      <div>
        <ProductInfo
          countComments={0}
          currentPrice={0}
          product={Object {}}
        />
        <ProductSizes
          currentSize={
            Object {
              "_id": "",
            }
          }
          handleSizeChange={[Function]}
          sizeIsNotSelectedError={false}
        />
        <div>
          <ProductSubmit
            disabled={true}
            product={Object {}}
            productToSend={
              Object {
                "options": Object {
                  "size": Object {
                    "_id": "",
                  },
                },
                "price": 0,
                "product": Object {
                  "_id": "",
                },
                "quantity": 1,
              }
            }
            setSizeIsNotSelectedError={[Function]}
          />
        </div>
      </div>
    </div>
  </WithStyles(ForwardRef(Card))>
</div>
`;
